src/Algebra/Construct/Initial.agda

```diff
@@ -0,0 +1,92 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Instances of algebraic structures where the carrier is ⊥.
+-- In mathematics, this is usually called 0.
+--
+-- From monoids up, these are zero-objects – i.e, the initial
+-- object is the terminal object in the relevant category.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Level using (Level)
+
+module Algebra.Construct.Initial {c ℓ : Level} where
+
+open import Algebra.Bundles
+  using (Magma; Semigroup; Band)
+open import Algebra.Bundles.Raw
+  using (RawMagma)
+open import Algebra.Core using (Op₂)
+open import Algebra.Definitions using (Congruent₂)
+open import Algebra.Structures using (IsMagma; IsSemigroup; IsBand)
+open import Data.Empty.Polymorphic
+open import Relation.Binary using (Rel; Reflexive; Symmetric; Transitive; IsEquivalence)
+
+
+------------------------------------------------------------------------
+-- re-export those algebras which are also terminal
+
+open import Algebra.Construct.Terminal {c} {ℓ} public
+  hiding (rawMagma; magma; semigroup; band)
+
+------------------------------------------------------------------------
+-- gather all the functionality in one place
+
+private module ℤero where
+
+  infixl 7 _∙_
+  infix  4 _≈_
+  Carrier : Set c
+  _≈_     : Rel Carrier ℓ
+  _∙_     : Op₂ Carrier
+
+  Carrier = ⊥
+  _≈_ ()
+  _∙_ ()
+
+  refl : Reflexive _≈_
+  refl {x = ()}
+
+  sym : Symmetric _≈_
+  sym {x = ()}
+
+  trans : Transitive _≈_
+  trans {i = ()}
+
+  isEquivalence : IsEquivalence _≈_
+  isEquivalence = record { refl = refl ; sym = sym ; trans = trans }
+
+  ∙-cong : Congruent₂ _≈_ _∙_
+  ∙-cong {x = ()}
+
+  isMagma : IsMagma _≈_ _∙_
+  isMagma = record { isEquivalence = isEquivalence ; ∙-cong = ∙-cong }
+
+  isSemigroup : IsSemigroup _≈_ _∙_
+  isSemigroup = record { isMagma = isMagma ; assoc = λ () }
+
+  isBand : IsBand _≈_ _∙_
+  isBand = record { isSemigroup = isSemigroup ; idem = λ () }
+
+open ℤero
+
+------------------------------------------------------------------------
+-- Raw bundles
+
+rawMagma : RawMagma c ℓ
+rawMagma = record { Carrier = Carrier ; _≈_ = _≈_ ; _∙_ = _∙_ }
+
+------------------------------------------------------------------------
+-- Bundles
+
+magma : Magma c ℓ
+magma = record { Carrier = Carrier ; _≈_ = _≈_ ; _∙_ = _∙_ ; isMagma = isMagma }
```

Comment:
You only need `magma = record { isMagma = isMagma }` here and Agda will infer the rest. Ditto elsewhere.

Reply:
For some reason (as with Jacques' comments) I had tried this, but not managed to get it to work. I'll try again.

Reply:
For `Initial`, this seemed to work. But I don't understand what to do for `Terminal`. What I really wanted to achieve was to define a single record `one` packaging the `Carrier` and ` _≈_` in module `𝕆ne` for which all the definitions in this file would consist of writing `... = one` or `... = record { one }`, but this didn't seem to work. But the previous version of these definitions, with repetitions of `Carrier = ⊤ ; _≈_  = λ _ _ → ⊤` seemed even worse!

Reply:
I think this will indeed only work for `Initial`. What's going on is probably quite subtle, and having to do with Agda being quite happy to eagerly unify things when it's clear there's only one, but not nearly as keen to fill in `bottom` when it should be equally clear that something is impossible.

Which is actually sensible: a priori, there's no reason that witnesses of impossibility are unique. In fact, being too eager to match on `bottom` would have made it so that `Set` would not be a rig category because the coherence law between the two ways of doing `0*0=0` would not hold!!

Reply:
No I think it's much simpler than that. It's that Agda can simply infer `_≈_` and `_∙_` (and hence `Carrier`) are from something of type `IsMagma _≈_ _∙_`